### PR TITLE
feat: localization evaluator

### PR DESCRIPTION
**Pull request by angry-crab**
Signed-off-by: Xinyu Wang <xinyu.wang@tier4.jp>

## Description

Localization evaluator (tested with Kitti raw dataset)

## Related links

[Evaluate performance of localization pipeline](https://github.com/autowarefoundation/autoware.universe/issues/602)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by angry-crab** [comment]
Comments will be removed after review.

File: localization/kitti_evaluator/package.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>kitti_evaluator</name>
```
Comment (mitsudome-r) [inline localization/kitti_evaluator/package.xml:4]: If this tool is just comparing localization output and ground truth values, then I think we can use more generic name such as localization_evaluator

File: localization/kitti_evaluator/include/kitti_evaluator/kitti_evaluator_core.hpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    Eigen::Affine3d & vehicle_trans, Eigen::Affine3d & groud_truth_trans,
+    Eigen::Affine3d & error_trans);
+  void publishError(Eigen::Affine3d & error_trans);
+
+  // rclcpp::Subscription<sensor_msgs::msg::NavSatFix>::SharedPtr nav_sat_fix_sub_;
+  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr vehicle_pose_sub_;
+  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr groud_truth_pose_sub_;
+  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pose_error_pub_;
+
+  Eigen::Affine3d last_vehicle_trans;
```
Comment (mitsudome-r) [inline localization/kitti_evaluator/include/kitti_evaluator/kitti_evaluator_core.hpp:58]: There should be underscore at the end of the variable name to keep naming convention consistent with other member variables.
Same goes to the rest of the variables

File: localization/kitti_evaluator/src/kitti_evaluator_core.cpp [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+      msgToMatrix(prev_pose, prev_trans);
+    }
+    Eigen::Affine3d post_trans;
+    rclcpp::Time post_time(post_pose->header.stamp);
+    msgToMatrix(post_pose, post_trans);
+
+    rclcpp::Time fake_time = curr_time - rclcpp::Duration::from_seconds(delta);
+    double t = interpolation::getTimeCoeffients(fake_time, prev_time, post_time);
+    if (t > 1 || t < 0) {
+      RCLCPP_INFO(get_logger(), "????????????");
```
Comment (mitsudome-r) [inline localization/kitti_evaluator/src/kitti_evaluator_core.cpp:143]: This should have proper output message. We might want to consider skipping the rest of the process if there is something wrong with the interpolation time.

**Comment by mitsudome-r** [comment]
I have talked with @angry-crab to refactor the code according to the diagram in the pdf.
[Uploading Localization Evaluator.pdf…]()

**Comment by mitsudome-r** [comment]
The code looks good to me. Maybe we should have the following in the readme:
* How to convert KITTI dataset into a rosbag.
* Recommended way to use this tool (e.g., with plotjuggler)

**Comment by angry-crab** [comment]
> The code looks good to me. Maybe we should have the following in the readme:
> 
> * How to convert KITTI dataset into a rosbag.
> * Recommended way to use this tool (e.g., with plotjuggler)
>   It can be done with another PR if that sounds suitable.

README is updated to include more detailed instructions and recommendation usage.

**Review by mitsudome-r — APPROVED** [review]

File: localization/localization_evaluator/README.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+`ros2 bag play PATH_TO_ROS2BAG/ros2_bag.db3 --qos-profile-overrides-path qos.yaml -d 3`
+
+- New Terminal 4
+
+`source ~/autoware/install/setup.bash`
+`ros2 service call /rosbag2_player/toggle_paused rosbag2_interfaces/srv/TogglePaused`
+
+- Go to Rviz of Terminal 1, set an intial pose, when Terminal 1 output `Alignment called`, go to Terminal 4 and rerun the last command. Then you are good to go.
+
+Check the link below for a video demonstration. [Sample Video](https://github.com/orgs/autowarefoundation/discussions/334)
```
Comment (mitsudome-r) [inline localization/localization_evaluator/README.md:94]:
```suggestion
Check the link below for a video demonstration. [Sample Video](https://github.com/autowarefoundation/autoware/discussions/2618)
```

**Comment by angry-crab** [comment]
@mitsudome-r I reverted some weird commits. Please check again. If everything looks good, I'll merge this PR. Thanks.

**Comment by xmfcx** [comment]
@angry-crab are there any updates on this? It is approved but CI build seems to be failing.

**Comment by angry-crab** [comment]
> @angry-crab are there any updates on this? It is approved but CI build seems to be failing.

This PR will be closed because there is already a package which does the same thing. I'll create another PR to merge some features into the existing package.

**Pull request closed without merging**